Mathlib/Data/Nat/Bitwise.lean

```diff
@@ -326,6 +344,61 @@ theorem bitwise_swap {f : Bool → Bool → Bool} :
     simp [bitwise_of_ne_zero, ih]
 #align nat.bitwise_swap Nat.bitwise_swap
 
+theorem testBit_two_pow_mul_add {n b w i} (h: i < w) :
+    Nat.testBit (2 ^ w * b + n) i = Nat.testBit n i := by
+  simp only [testBit, shiftRight_eq_div_pow, bodd_eq_bodd_iff]
+  rw [Nat.add_div_of_dvd_right (by simp [Dvd.dvd.mul_right, pow_dvd_pow, le_of_lt h]), add_mod]
+  have h1: (2 ^ w / 2 ^ i) % 2 = 0 := by
+    rw [← Nat.dvd_iff_mod_eq_zero]
+    use 2 ^ (w - (i + 1))
+    rw [Nat.pow_div (by linarith) (by decide), mul_comm, ← pow_succ 2 _, succ_eq_add_one]
+    simp [← Nat.sub_add_comm, succ_le_of_lt h]
+  simp [mul_comm, Nat.mul_div_assoc b (pow_dvd_pow 2 (le_of_lt h)), mul_mod, h1]
+
+theorem testBit_two_pow_mul_toNat_add {n w b} (h: n < 2 ^ w) :
+    testBit (2 ^ w * b.toNat + n) w = b := by
+  simp only [testBit, shiftRight_eq_div_pow]
+  rw [Nat.add_div_of_dvd_right (Dvd.intro _ rfl), Nat.div_eq_of_lt h, add_zero]
+  cases' b <;> simp
+
+/-- Generic method to create a natural number by appending bits tail-recursively.
+It takes a boolean function `f` on each bit the number of bits `i`.  It is
+almost always specialized  `i = w`; the length of the binary representation.
+This is an alternative to using `List`. It will be used for bitadd, bitneg, bitmul etc.-/
+def ofBits (f : Nat → Bool) : Nat → Nat :=
```

Comment:
I'd expect this to be more convenient as 
```suggestion
def ofBits {i : Nat} (f : Fin n → Bool) : Nat :=
```
that is, make `f` dependently typed. Otherwise you need a congruence lemma that says `f j` is irrelevant unless `j < i`.

Reply:
I tried but don't yet see what to do here. Someone else has an idea?

Reply:
I mean something like
```
def ofBits {n : Nat} (f : Fin n → Bool) : Nat :=
  go f 0
  where
    /-- A helper method where `z` the starting point.
    Note that `ofBits.go f z = 2 ^ n * z + ofBits f i` which we prove next. -/
    go {n : Nat} (f : Fin n → Bool) (z : Nat) : Nat :=
    match n with
    | 0 => z
    | _ + 1 => go (f ∘ Fin.castSucc) (z.bit (f (.last _)))
```
I'll push something to that effect shortly.

Reply:
What is the motivation behind defining it in terms of `Fin`? I originally defined it this way to avoid lists (which is equivalent to this new definition). This was mainly because our proofs of bit-blasting theorems all follow this strong induction format where we descend to smaller binary representations (removing elements is painful to do with lists).

Also, I'm trying to prove `testBit_ofBits` with this new definition (which was a simple 5-line strong inductive proof before) but I can't seem to generalize over `i` anymore. Issues like this make me favor the `Nat` definition over the `Fin` definition (dependent types are harder to work with). Am I missing something?

Reply:
In some sense, the `Fin` design is more precise, since it's now obvious that only the values of `f i` where `i < n` matter. The previous design required garbage values `f j` for all `j >= n`, and, like Eric mentioned, some congruence lemma stating that these garbage values don't influence the output.

That said,
> I'd expect this to be more convenient
So far, the proofs that were broken are considerable less convenient. 
I've tried to play around with different `Fin` recursion principles to regain the proofs as straightforward inductions, but have not gotten that to work.

@eric-wieser: What do you think of keeping the old `Nat -> Bool`-based version as a (private) implementation detail, and have the `Fin n -> Bool` version just wrap that one (by extending the function with canonical garbage values). That should be easily implemented over the previously working version.

Reply:
I like that idea.